nifi-nar-bundles/nifi-marklogic-bundle/nifi-marklogic-processors/src/main/java/com/marklogic/nifi/processor/PutMarkLogic.java

```diff
@@ -0,0 +1,396 @@
+        .name("Permissions")
+        .displayName("Permissions")
+        .defaultValue("rest-reader,read,rest-writer,update")
+        .description("Comma-delimited sequence of permissions - role1, capability1, role2, " +
+            "capability2 - to add to each document")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .build();
+
+    public static final PropertyDescriptor TEMPORAL_COLLECTION = new PropertyDescriptor.Builder()
+        .name("Temporal collection")
+        .displayName("Temporal collection")
+        .description("The temporal collection to use for a temporal document insert")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .build();
+
+    public static final PropertyDescriptor TRANSFORM = new PropertyDescriptor.Builder()
+        .name("Server transform")
+        .displayName("Server transform")
+        .description("The name of REST server transform to apply to every document as it's" +
+            " written")
+        .addValidator(Validator.VALID)
+        .required(false)
+        .build();
+
+    public static final PropertyDescriptor URI_ATTRIBUTE_NAME = new PropertyDescriptor.Builder()
+        .name("URI attribute name")
+        .displayName("URI attribute name")
+        .defaultValue("uuid")
+        .required(true)
+        .description("The name of the FlowFile attribute whose value will be used as the URI")
+        .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+        .build();
+
+    public static final PropertyDescriptor URI_PREFIX = new PropertyDescriptor.Builder()
+        .name("URI prefix")
+        .displayName("URI prefix")
+        .description("The prefix to prepend to each URI")
+        .required(false)
+        .addValidator(Validator.VALID)
+        .build();
+
+    public static final PropertyDescriptor URI_SUFFIX = new PropertyDescriptor.Builder()
+        .name("URI suffix")
+        .displayName("URI suffix")
+        .description("The suffix to append to each URI")
+        .required(false)
+        .addValidator(Validator.VALID)
+        .build();
+
+    protected static final Relationship SUCCESS = new Relationship.Builder()
+        .name("SUCCESS")
+        .description("All FlowFiles that are successfully written to MarkLogic are routed to the " +
+            "success relationship for future processing.")
+        .build();
+
+    protected static final Relationship FAILURE = new Relationship.Builder()
+        .name("FAILURE")
+        .description("All FlowFiles that failed to be written to MarkLogic are routed to the " +
+            "failure relationship for future processing.")
+        .build();
+
+    private DataMovementManager dataMovementManager;
+    private WriteBatcher writeBatcher;
+    // If no FlowFile exists when this processor is triggered, this variable determines whether or not a call is made to
+    // flush the WriteBatcher
+    private boolean shouldFlushIfEmpty = true;
```

Comment:
It appears this member variable could benefit from being 'volatile'.  This is true for the writeBatcher and possibly dataMovementManager as well.